File: Recent_Search/v1/Python/recent-search.py
import json
import os
import urllib.parse
import requests
from requests.auth import AuthBase

from dotenv import load_dotenv
load_dotenv(verbose=True)  # Throws error if it can't find .env file

# Retrieves credential information from the '.env' file
BEARER_TOKEN = os.getenv("TWITTER_BEARER_TOKEN")

SEARCH_URL = "https://api.twitter.com/labs/1/tweets/search"
QUERY = "(snow OR hail OR rain) (Colorado OR #COWX) has:images"
MAX_REQUESTS = 3
OPTIONS = "&format=compact" #Labs v1 only.

def get_tweets(auth, query):

    url = f"{SEARCH_URL}?query={QUERY}{OPTIONS}" #{urllib.parse.quote(REQUEST_PARAMETERS)}"
    response = requests.get(url, auth=auth, headers = headers)

    if response.status_code is not 200:
        #raise Exception(f"Error with request (HTTP error code: {response.status_code} - {response.reason}")
        print (f"Error with request (HTTP error code: {response.status_code} - {response.reason}")

    return response
    
#-----------------------------------------------------------------------------------------------------------------------
if __name__ == "__main__":

  query = urllib.parse.quote(QUERY)
  bearer_token = BEARER_TOKEN

  #As we page through results, we will be counting these: 
  request_count = 0
  tweet_count = 0

  while True:
      #loop body
      request_count += 1
      response = get_tweets(bearer_token, query)
      parsed = json.loads(response.text)
      print (json.dumps(parsed, indent=2, sort_keys=True))

      try:
          next_token  = parsed['meta']['next_token']
      except KeyError:
          next_token = None

      try:
          tweet_count  += parsed['meta']['result_count']
      except KeyError:
          pass

      if (next_token is None or request_count == MAX_REQUESTS): break

  print(f"Made {request_count} requests and received {tweet_count} Tweets...")

File: Recent_Search/v1/Ruby/recent-search.rb
require "net/https"
require "uri"
require 'base64'
require 'json'
require 'dotenv/load'

@query = '(Twitter Labs search) OR from:TwitterDev OR from:SnowBotDev OR from:DailyNASA'
@max_requests = 3

SEARCH_URL = "https://api.twitter.com/labs/1/tweets/search"

def make_request(bearer_token, query)

  uri = URI(SEARCH_URL)

  options = {}
  options['query'] = query
  options['max_results'] = 10
 
  uri.query = URI.encode_www_form(options)

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  request = Net::HTTP::Get.new(uri.request_uri)
  request['Authorization'] = "Bearer #{bearer_token}"
  request['User-Agent'] = "RecentSearchQuickStartRuby"

  response = http.request(request)

  return response
end

if __FILE__ == $0  #This script code is executed when running this file.

  bearer_token = ENV['TWITTER_BEARER_TOKEN']

  request_count = 0
  tweet_count = 0

  loop do
    request_count += 1
    response = make_request(bearer_token, @query)
    response_json = JSON.parse(response.body)
    puts JSON.pretty_generate(response_json)

    next_token = response_json['meta']['next_token']
    tweet_count += response_json['meta']['result_count'].to_i

    break if next_token.nil? or request_count == @max_requests
  end

  puts "Made #{request_count} requests and received #{tweet_count} Tweets."

end
